refactor: Subplot layout adjustments

miroslaavi opened the pull request:
Fixes #179 

## What does this implement/fix? Explain your changes.
This pull request introduces a new function _set_subplot_default_kwargs to dynamically set default plot sizes for subplots, enhancing the readability and aesthetics of plots in the plotting module. This function is now utilized in plot_panel, plot_forecasts, and plot_backtests to ensure consistent and optimal sizing of plots, especially when the user hasn't specified height and width in kwargs.

Additionally, this PR includes a fix for the #179 for the subplot functions plot_forecast, plot_panel, and plot_backtests.

## Any other comments?
Also, added a logic if n_series is greater than the max unique entities, then use max unique entities instead for plotting to avoid an error.

miroslaavi commented:
@baggiponte tagging you here as the first fixes are ready for review, looks like I am not able to add a reviewer yet. :)

baggiponte commented:
Sorry for the late reply. Still under pressure, will try to review this ASAP.

baggiponte commented:
Starting review, might still take a few days. First of all: excellent job, high quality comments, names and abstractions over repetitive stuff. Will leave some comments here and there.

I would kindly ask you to add some tests for the functions you added. Feel free to place them under a new `tests/test_plotting.py`.

baggiponte commented:
Kudos for removing `np.repeat` usage - that thing bugged me for waaay to much, but I never had the time to clean that up.

We do that "iterate over entities" at least three times in the module. Would you mind making that a standalone function and add tests? Of course, available to help. You don't need to test the tests, but make sure that the `divmod` vs `np.repeat` behaves the same.

FBruzzesi reviewed (commented):
I am also giving a quick review, as @baggiponte called me in (I feel like a reviewer-ad-honorem 😂).

I left a minor suggestion - as such function has no facing api, but still good for standardization.

As @baggiponte mentioned, I aswell think there are other tiny code parts that can be refactored into a utility function.

Regarding unit tests, sadly I never had to test plotting functionalities.

miroslaavi commented:
Sounds great, thanks both for taking the time. I will look into these over the weekend!

baggiponte commented:
Thanks! One more thing: I updated main, could you rebase your PR? We'd prefer that to merging master, to keep a simpler commit history.

miroslaavi commented:
@baggiponte @FBruzzesi, I have now tried to module some of the repetitive code to own functions, but not sure what is the best way to deal with the small differences in plotting (naming, colors etc.). I've now added plot_params dict prior calling the "_add_scatter_traces_to_subplots".. maybe I just made it more complicated than it was haha :D 

I've tried with multiple different len series and columns and plotting looks to work well now with plot_panel, plot_forecast, and plot_backtests. I also added the tests to the test_plotting.py, have a look!

miroslaavi commented:
To add, I also tested some of the other plotting functions and noticed that the plots using plotly express doesn't work no more and leads to AssertionError. Not quite sure what causes it but most likely something changed either in plotly or polars. In order to make it work now, I rewrote the plot_entities to go.Figure instead, and for other plot functions made a quick fix and converting the polars DF to pandas prior plotting with .to_pandas()

miroslaavi commented:
> Thanks! One more thing: I updated main, could you rebase your PR? We'd prefer that to merging master, to keep a simpler commit history.

Cool, never done that prior as mostly used version history for my own projects that I was working alone. Also, noticed that my commit messages are not very constant or following a clear structure, so I'll check some best practices for that too. :)

Now it is like 10 commits for these changes, is it preferred to keep that way or is it ideal to squeeze in one commit for the merge?

baggiponte commented:
Ciao Miro, sorry for the late reply. Have a workshop tomorrow at EPFL. Should have time to review afterwards.

miroslaavi commented:
All good and take your time, have a great workshop! :)

baggiponte commented on functime/plotting.py:
```diff
@@ -23,6 +22,171 @@ def _remove_legend_duplicates(fig: go.Figure) -> go.Figure:
+    return n_rows
+
+
+def _get_subplot_grid_position(i, n_cols):
```
Please add type hints 😊

baggiponte commented on functime/plotting.py:
```diff
@@ -23,6 +22,171 @@ def _remove_legend_duplicates(fig: go.Figure) -> go.Figure:
+    return row, col
+
+
+def _prepare_data_for_subplots(y: Union[pl.DataFrame, pl.LazyFrame], n_series: int, last_n: int, seed: Union[int, None] = None):
```
Please add return types :)

baggiponte commented on functime/plotting.py:
```diff
@@ -23,6 +22,171 @@ def _remove_legend_duplicates(fig: go.Figure) -> go.Figure:
+
+    Returns
+    -------
+    Tuple
```
Use `Tuple[<Dtype>, <Dtype>, <dtype>]` with appropriate types

baggiponte reviewed (changes requested):
Thanks for the great work. One last thing: run the ruff and isort formatter and add the remaining type hints, then we can merge!

miroslaavi commented:
@baggiponte, thanks! This was quite many changes in one PR, but going forward will raise issues and focus on smaller chunks. This is good to go from my side now.

baggiponte commented:
Amazing, will merge soon. Indeed it's a big PR but we requested most of the changes. Thank you very much for your contribution and patience 🫡
